File: lib/commands/remove.js
const fs = require("fs");
const path = require("path");
const readline = require("readline");
const { gitInBare } = require("../git");
const { loadConfig } = require("../config");

async function remove(branch, opts) {
  const { root } = loadConfig();
  const worktreePath = path.join(root, branch);
  const forceFlag = opts.force ? " --force" : "";

  // Resolve the actual git branch name from the worktree before removing it
  // (may differ from the directory name, e.g. prs/123 -> feature-xyz)
  let branchName = branch;
  try {
    const dotGit = fs.readFileSync(path.join(worktreePath, ".git"), "utf-8")
      .replace("gitdir: ", "").trim();
    const head = fs.readFileSync(path.join(dotGit, "HEAD"), "utf-8").trim();
    const match = head.match(/^ref: refs\/heads\/(.+)$/);
    if (match) branchName = match[1];
  } catch {}

  // Remove worktree
  console.log(`Removing worktree "${branch}"...`);
  try {
    gitInBare(`worktree remove ${worktreePath}${forceFlag}`, root);
  } catch (err) {
    console.error(`Failed to remove worktree: ${err.message}`);
    if (!opts.force) {
      console.error("Use --force to remove a dirty worktree.");
    }
    process.exit(1);
  }

  // Ask whether to delete the branch
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const answer = await new Promise((resolve) => {
    rl.question(`Delete branch "${branchName}" as well? (y/N): `, resolve);
  });
  rl.close();

  if (answer.trim().toLowerCase() === "y") {
    try {
      gitInBare(`branch -d ${branchName}`, root);
      console.log(`Branch "${branchName}" deleted.`);
    } catch {
      console.error(
        `Could not delete branch "${branchName}". It may not be fully merged.`
      );
      console.error(`Use \`git branch -D ${branchName}\` to force delete.`);
    }
  }

  console.log("Done.");
}

module.exports = remove;

File: internal/cmd/remove.go
package cmd

import (
	"fmt"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
)

// resolveWorktreeFromDir finds the worktree that contains dir (either an exact
// match or a parent of dir).
func resolveWorktreeFromDir(dir string, entries []WorktreeEntry) (*WorktreeEntry, error) {
	for _, e := range entries {
		if dir == e.Dir || strings.HasPrefix(dir, e.Dir+string(filepath.Separator)) {
			return &e, nil
		}
	}
	return nil, fmt.Errorf("current directory is not inside a worktree")
}

func RemoveCmd(d *Deps) *cobra.Command {
	var force bool
	cmd := &cobra.Command{
		Use:   "remove [name]",
		Short: "Remove a worktree and optionally delete the branch",
		Long:  "Remove a worktree and optionally delete the branch.\nIf no name is given, removes the worktree for the current directory.",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			root, _, err := d.Config.LoadConfig()
			if err != nil {
				return err
			}

			// Resolve the worktree from the list (supports branch name, path, substring)
			output, err := d.Git.GitInBare("worktree list", root)
			if err != nil {
				return err
			}
			entries := ParseWorktreeList(output)

			var entry *WorktreeEntry
			if len(args) == 1 {
				entry, err = ResolveWorktree(args[0], root, entries)
				if err != nil {
					return err
				}
			} else {
				// No arg: detect current worktree from cwd
				cwd, err := d.Getwd()
				if err != nil {
					return fmt.Errorf("could not determine current directory: %w", err)
				}
				entry, err = resolveWorktreeFromDir(cwd, entries)
				if err != nil {
					return err
				}
			}

			worktreePath := entry.Dir
			branchName := entry.Branch

			// Remove worktree
			fmt.Fprintf(d.Stdout, "Removing worktree %q...\n", branchName)
			removeArgs := "worktree remove " + worktreePath
			if force {
				removeArgs += " --force"
			}
			if _, err := d.Git.GitInBare(removeArgs, root); err != nil {
				// If not already forcing and the error is about dirty files, prompt to force
				if !force && strings.Contains(err.Error(), "modified or untracked") {
					answer := d.Prompt.Prompt("Worktree has modified or untracked files. Force remove? (y/N)", "")
					if strings.ToLower(strings.TrimSpace(answer)) == "y" {
						if _, err := d.Git.GitInBare("worktree remove "+worktreePath+" --force", root); err != nil {
							return fmt.Errorf("force removal failed: %w", err)
						}
					} else {
						return fmt.Errorf("worktree removal cancelled")
					}
				} else {
					return fmt.Errorf("failed to remove worktree: %w", err)
				}
			}

			// Ask whether to delete the branch
			answer := d.Prompt.Prompt(fmt.Sprintf("Delete branch %q as well? (y/N)", branchName), "")

			if strings.ToLower(strings.TrimSpace(answer)) == "y" {
				// Use -D (force) for PR worktrees since the branch will be recreated
				// from remote by `wt pr`. Use -d (safe) for regular worktrees.
				rel, _ := filepath.Rel(root, worktreePath)
				isPr := strings.HasPrefix(rel, "prs/") || strings.HasPrefix(rel, "prs"+string(filepath.Separator))
				deleteFlag := "-d"
				if isPr {
					deleteFlag = "-D"
				}
				if _, err := d.Git.GitInBare(fmt.Sprintf("branch %s %s", deleteFlag, branchName), root); err != nil {
					fmt.Fprintf(d.Stderr, "Could not delete branch %q. It may not be fully merged.\n", branchName)
					fmt.Fprintf(d.Stderr, "Use `git branch -D %s` to force delete.\n", branchName)
				} else {
					fmt.Fprintf(d.Stdout, "Branch %q deleted.\n", branchName)
				}
			}

			fmt.Fprintln(d.Stdout, "Done.")
			return nil
		},
	}
	cmd.Flags().BoolVar(&force, "force", false, "force removal even if worktree is dirty")
	return cmd
}
